app/services/subprocess_runner.py

```
@@ -427,6 +527,17 @@ async def _cancel_watcher():
             buffer = ""
             output_lines: list[str] = []
             stall_error: str | None = None
+            abandoned_error: str | None = None
+            # Native stdout parsing is the preferred signal; the size-growth
+            # fallback below only speaks while this stays False.
+            saw_native_progress = False
+            ratio = size_ratio_for(mode)
+            expected_size = 0
+            if ratio:
+                try:
+                    expected_size = max(1, int(os.path.getsize(input_path) * ratio))
```

Ruff (0.16.1)

[warning] Async functions should not use os.path methods, use trio.Path or anyio.path

(ASYNC240)